Support for any type of custom route builder.

jeduden opened the pull request:
Example for DialogRoute

slovnicki commented:
@jeduden thanks for the PR! I will be reviewing it in the evening

jeduden commented:
@slovnicki do you think this approach looks fine in principle ? If yes, i ll see if i can try to add a unit-test.

slovnicki commented:
@jeduden I would say that this is definitely a step in the right direction. We would like to handle all kind of `Route`s (not just `PageRoute`s). I have a concern then that it's not correct to also use `BeamPage` for the dialog that is "pageless", but I'll have to run the example to see everything in more detail.

Nevertheless, your PR gives an insight into what we might be missing and how to handle it more cleanly within the beamer package itself. We might want to introduce some new "navigation objects" beside `BeamPage`s. This is definitely something to focus on, nice job!

I'll leave more comments when I start a review :slightly_smiling_face:

jeduden commented:
@slovnicki ok, great! i ll hold off then until the full review.

slovnicki commented:
@jeduden I'm back from a vacation and I'll definitely be looking into this during the week

slovnicki commented on package/lib/src/beam_page.dart:
```diff
@@ -26,6 +26,7 @@ class BeamPage extends Page {
     this.popToNamed,
     this.type = BeamPageType.material,
     this.pageRouteBuilder,
+    this.routeBuilder,
```
I'm think we should then remove `pageRouteBuilder` and use only `routeBuilder` for all. (`PageRoute` is a subclass of `Route`)

slovnicki replied:
oh... And when you remove `pagRouteBuilder`, you'll have to rename it to `routeBuilder` in all tests and examples where it's used (probably in no examples, but tests yes).

slovnicki commented on package/lib/src/beam_page.dart:
```diff
@@ -105,6 +106,13 @@ class BeamPage extends Page {
   final PageRoute Function(RouteSettings settings, Widget child)?
       pageRouteBuilder;
```
again, we should then remove this

slovnicki commented on package/lib/src/beam_page.dart:
```diff
@@ -116,6 +124,9 @@ class BeamPage extends Page {
     if (pageRouteBuilder != null) {
       return pageRouteBuilder!(this, child);
```
and of course, we remove this as well

slovnicki commented on examples/books_dialogroute/lib/main.dart:
```diff
@@ -0,0 +1,152 @@
+                      context: context,
+                      builder: (context) => child,
+                      settings: settings),
+              key: ValueKey('book-buy-${state.pathParameters['bookId']}'),
```
it's a bit unexpected that `BooketailsScreen` pops when we click outside of dialog to dismiss it. This is due to the `BeamPage.defaultOnPopPage` that will remove the `bookId` path segment.

We need a custom `onPopPage` like this:
```dart
onPopPage: (context, delegate, page) {
  delegate.currentBeamLocation.update(
    (state) => state.copyWith(queryParameters: {}),
  );
  return true;
},
```

slovnicki commented:
@jeduden I left you some reviews. Overall, this seems good to me. Unfortunately, we'll have to make a breaking change of using only `routeBuilder` and removing `pageRouteBuilder` (effectively, users who use it will just rename the property).

When you push new commits to resolve the reviews, I will review again and merge

jeduden commented:
Awesome, will work on this hopefully this week.

jeduden commented:
@slovnicki I have updated the PR.
- removed the `pageRouteBuilder` property
- updated the documentation and described the child parameter the route builder received.
- updated the unittests. I also changed the unittest setup slightly:  it tests that the `child`  the route builder receives is in fact the child of the beam page.
- none of the other examples used `pageRouteBuilder`  - am I missing something?
- added a breaking change entry to the changelog.

slovnicki reviewed (approved):
Nice work @jeduden!
Thanks again for this PR.
